### ISSUE TITLE
Make the leaflet container a programmatically identifiable region

### ISSUE THREAD

**Issue by Malvoz** [Leaflet/Leaflet#7193]
It is easy to determine that a Leaflet map is an interactive map that is embedded on a web page when we see it. This may not be the case for visually impaired users who rely on their screen reader. Therefore I propose making the `.leaflet-container` a programmatically identifiable region by 1) adding a landmark role ([`role="region"`](https://www.w3.org/TR/wai-aria/#region)), and 2) adding an [accessible name](https://www.w3.org/TR/accname-1.1/#dfn-accessible-name) that describes the purpose of the element (e.g. `aria-label="Interactive map"`).

[Landmark roles](https://www.w3.org/TR/wai-aria/#landmark_roles) are special types of roles in that they allow assistive technologies to include the landmarks in a page summary, and also enables landmark-based keyboard navigation.

This is essentially implementing the WCAG 2.1 [technique ARIA11](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA11) which satisfies Success Criterion [1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships), [1.3.6: Identify Purpose](https://www.w3.org/WAI/WCAG21/Understanding/identify-purpose) and [2.4.1: Bypass Blocks](https://www.w3.org/WAI/WCAG21/Understanding/bypass-blocks) for the leaflet container.

Related resources:
- https://www.scottohara.me/blog/2018/03/03/landmarks.html
- https://www.w3.org/TR/wai-aria-practices/examples/landmarks/
- https://www.w3.org/TR/wai-aria-practices/examples/landmarks/region.html

**Comment by Falke-Design** [comment]
I would say `aria-label="Map"` would be make sense but like I understand it, with defining `role='region'` we would say that the Leaflet map is one of th main parts of the website, but this is not everytime correct. Leaflet is also often used to show an address in a contact form like [here](https://preview.colorlib.com/#meetup). And sometimes the map is not accessible and then it would be wrong to say it is `role='region'`. I think this needs to be done by the developers self.

**Comment by Malvoz** [comment]
I think independently of whether the map is part of a form or other semantic element an appropriate default `role` would be useful to provide additional context. Currently the components of the map are seemingly not part of any particular structure (other than the root document, of course), see the current Accessibility Tree:

![zoom-button-context](https://user-images.githubusercontent.com/26493779/142779754-2b9990aa-db65-4291-b05d-182130711824.png)

But I agree that `role="region"` is giving the maps too much importance so to speak, I think `group` is a better role for this purpose:

![zoom-button-context-after](https://user-images.githubusercontent.com/26493779/142779874-46a16472-0d2b-41c2-85b6-c4d7fc681713.png)

**Comment by Malvoz** [comment]
[`aria-roledescription`](https://www.w3.org/TR/wai-aria-1.1/#aria-roledescription) makes even more sense than `aria-label` as it overrides the `role` (basically a custom role), _and_ allows developers to uniquely describe each map using `aria-label`.

For example, if the default is:

```html
<div id="map" role="group" aria-roledescription="map"></div>
```

A developer can add an `aria-label="My country"` without overriding "map"; a screen reader would announce "My country, map" in similar fashion to e.g. "Confirm, button" for `<button>Confirm</button>`.

**Comment by nfreear** [comment]
Hi,
:+1: I'm very happy to see this issue raised. A few points:

* To stress what was posted originally, [interactive elements must have an appropriate role and label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Text_labels_and_names#interactive_elements_must_be_labeled), so that assistive technology users have a clue why and how they would interact with the element (in this case, so they can pan and zoom the map using arrow keys and +/- respectively.)

* Where the map container is keyboard focusable and interactive (the default), it does not seem appropriate to use the `group` role. A `group` is equivalent to a `<fieldset>`, it should not itself be interactive, and is expected to contain interactive widgets (and potentially other groups).

* Where the map container is NOT focusable and interactive (Leaflet `keyboard` option set to "false"?), whether it contains static markers or ones with pop-ups, controls or not, then use `role="group"`.

* Where the map container is focusable and interactive, only contains static markers, and does NOT contain interactive controls (Zoom in/out, attribution), then the `button` role may be closest [Case: "_semi-static_"].

* Where the map container is focusable and interactive, and contains interactive controls and markers with pop-ups and so on, then the `region` role may be closest (note, normally landmark roles are not focusable - an exception?) It could contain a `group` for markers, and a `role="toolbar"` for map controls. Is this the most typical case? [Case: "_interactive_"]

* `aria-roledescription` - In all cases, I agree that `aria-roledescription` is useful to make the role more specific. This could be set to "map", "interactive map", "static map" or similar. Important note: the value needs to be translated/localized by the developer implementing the map (unlike `role`, which is translated by the assistive technology).

* `aria-describedby` - where the map container is interactive, the user needs to know what keyboard shortcuts pap and zoom. This could be described in an element that is associated with `aria-describedby`. Note, this keyboard help element should not be hidden, as it is for all keyboard users.

Can we work out in code whether the map is _interactive_, _semi-static_ and so on? Count the number of controls, determine whether markers open pop-ups? Otherwise, do we need an `isInteractive` configuration option? Or, have a default `role` that can be configured directly, with guidance documentation?

I hope this helps move the discussion on.

What needs to happen to get a fix into Leaflet?

Thank you,

Nick
